### PR TITLE
Partial frontend caching of cluster responses (SCP-3284, SCP-3285, SCP-3286, SCP-2957)

### PR DESCRIPTION
**Pull request by devonbush**
SCP-3284, SCP-3285, SCP-3286, SCP-2957

Provides a transparent caching mechanism for calls to fetchCluster.  For each call, it attempts to cache three separate things from the response:
    cellsAndCoords: the coordinates and cell names of the cluster
    annotations: an array of annotations
    expression: an array of expression values
  
Prior to a new call, it checks the cache if any of those are available for the given params.  If so, it removes those fields from the list of fields to request from the server.  Once the response from the server comes back, it handles merging in all the cached fields into the response.

See the top of `plotDataCache.js` for full description.  

TO TEST:  This is pretty far-ranging in its impact.  So use https://docs.google.com/document/d/1CfpDVfmHJPvHISt-e1v90grKu9LrIrtEClxlRQJtFDs/edit, and make special effort to mix in very fast clicks (e.g. very quickly changing genes/annotations before some graphs finish loading)

TO TEST SCP-2957:  Load a study that has all zeros for a given gene expression (e.g. Male mouse brain, gene 'Adcy5').  Confirm that the coloring of the expression graph is no longer misleading.

### CODE REVIEW

File: app/controllers/api/v1/visualization/clusters_controller.rb
Comment (bistline) [inline app/controllers/api/v1/visualization/clusters_controller.rb:158]: This seems like it should be a constant for the controller as I expect it shouldn't change: `DEFAULT_DATA_FIELDS`

File: app/javascript/components/explore/plotDataCache.js
Comment (bistline) [inline app/javascript/components/explore/plotDataCache.js:226]: The structure of this method seems overly complicated... I'm wondering if it can be split up into some smaller helper methods to reduce the cyclomatic complexity.  I'm having a hard time reading it with all of the nested if/else logic.

File: app/javascript/components/explore/plotDataCache.js
Comment (bistline) [inline app/javascript/components/explore/plotDataCache.js:1]: Since there's a _lot_ going on here, it would be great if each method had standard docstrings with params/returns/etc.

File: app/javascript/components/visualization/ScatterPlot.js
Comment (bistline) [inline app/javascript/components/visualization/ScatterPlot.js:289]: Since the legend is auto-generated, and the sort order is determined by the input order of the traces (for group-based annotations), couldn't we just do a lexical sort of the traces by their name once we have all the data assembled?

File: app/lib/cluster_viz_service.rb
Comment (bistline) [inline app/lib/cluster_viz_service.rb:180]: This is starting to venture into the "too many parameters" range, even though they're keyword-based.  7 params seems like a lot.

File: app/lib/cluster_viz_service.rb
Comment (bistline) [inline app/lib/cluster_viz_service.rb:193]: I think this can all be simplified:
```
data_source = cluster
if annotation[:scope] == 'user'
  user_annotation = UserAnnotation.find(annotation[:id])
  subsample_annotation = user_annotation.formatted_annotation_identifier
  data_source = user_annotation
end
if include_coords
  x_array = data_source.concatenate_user_data_arrays('x', 'coordinates', subsample_threshold, subsample_annotation)
  y_array = data_source.concatenate_user_data_arrays('y', 'coordinates', subsample_threshold, subsample_annotation)
  z_array = data_source.concatenate_user_data_arrays('z', 'coordinates', subsample_threshold, subsample_annotation)
end
if include_cells || annotation[:scope] == 'study'
  # for study annotations, we have to grab the cluster cells to match with the study-wide annotation
  cells = data_source.concatenate_data_arrays('text', 'cells', subsample_threshold, subsample_annotation)
end
```

File: app/lib/cluster_viz_service.rb
Comment (bistline) [inline app/lib/cluster_viz_service.rb:249]: We still need to call `AnnotationVizService.sanitize_values_array` here so that we don't end up with blank values (that method interpolates `--Unspecified--` in for blank entries)

File: app/lib/expression_viz_service.rb
Comment (bistline) [inline app/lib/expression_viz_service.rb:115]: Same concern as above with not sanitizing this array.

**Review by bistline — COMMENTED** [review]
Great start!  Really liking where this is going.  I've put in some comments for areas that I think can be improved.

File: app/javascript/components/visualization/ScatterPlot.js
Comment (devonbush) [reply app/javascript/components/visualization/ScatterPlot.js:289]: I wish -- that's what we used to do.  Now we only send a single trace to plotly, and plotly's groupBy transformation handles splitting it -- so we don't have access to the subsequent generated traces.  So plotly determines the order based on the order the annotations are encountered in the *individual cells*, which would be extremely difficult to reorder.

File: app/javascript/components/explore/plotDataCache.js
Comment (devonbush) [reply app/javascript/components/explore/plotDataCache.js:1]: I added a bunch more function-level comments.  I'm hesitant to go full docstring, because there are a *lot* of small methods that all pass around the same params, and to me, the readability suffers if  `@param cluster (string) - the cluster name` gets repeated everywhere. But I'll defer to you and Eric on what you think

File: app/lib/cluster_viz_service.rb
Comment (devonbush) [reply app/lib/cluster_viz_service.rb:193]: NICE!

File: app/lib/cluster_viz_service.rb
Comment (devonbush) [reply app/lib/cluster_viz_service.rb:249]: yes, that gets taken care of just before the return

File: app/lib/expression_viz_service.rb
Comment (devonbush) [reply app/lib/expression_viz_service.rb:115]: The sanitization happens inside that utility function

File: app/lib/cluster_viz_service.rb
Comment (devonbush) [reply app/lib/cluster_viz_service.rb:180]: Yeah, not really sure what to do about it -- I could condense the include_* params into a single include object, but I think that makes it less clear.

**Review by eweitz — CHANGES_REQUESTED** [review]
Exciting stuff!  This seems like a huge optimization for single-gene search in Explore, and certain other use cases.

To help soundly quantify impact, we'll need to carefully code around `perfTime` logging as explained below.  I've prefixed those issues with "Blocker" for convenience.

Beyond those blocking functional problems, I suggest a bevy of non-blocking refinements for variable names, style, etc.

File: app/javascript/components/explore/ExploreDisplayTabs.js
Comment (devonbush) [reply app/javascript/components/explore/ExploreDisplayTabs.js]: I've changed to `createCache` to make it more verby

File: app/javascript/components/explore/plotDataCache.js
Comment (eweitz) [reply app/javascript/components/explore/plotDataCache.js]: With this name, I foresee maintainers likely wondering (and falsely assuming):

* Which cache, [server](https://github.com/broadinstitute/single_cell_portal_core/pull/1029)?  Nope, client.  (This is less concerning if we use a different prop name, like `cache: "client/server/none"`, for logging how/whether calls are cached.)

* Is the value giving me all the data that's in cache?  Instead, it's a Boolean.

File: app/javascript/components/explore/plotDataCache.js
Comment (eweitz) [reply app/javascript/components/explore/plotDataCache.js]: They note [classes can save memory better than objects](https://gist.github.com/indiesquidge/f8c486795d7dd455c0327ce7e0aa8c16#pitfalls-of-prototypal-inheritance).  Is that relevant here?

In any case, that gist is a great read -- thanks for sending it my way.  This approach works.

File: app/controllers/api/v1/visualization/clusters_controller.rb
Comment (eweitz) [reply app/controllers/api/v1/visualization/clusters_controller.rb]: Hmm, the discerning principle still seems murky -- it seems applicable to several other properties here that aren't prefixed with `userSpecified`.

File: app/lib/cluster_viz_service.rb
Comment (bistline) [inline app/lib/cluster_viz_service.rb:194]: I just tested this and realized it doesn't work, because the method in `UserAnnotation` is actually called `concatenate_user_data_arrays`, not `concatenate_data_arrays`.  So this either should revert back to what it was before, or that method in `UserAnnotation` should be renamed.  Either solution would be fine.

File: app/lib/cluster_viz_service.rb
Comment (bistline) [reply app/lib/cluster_viz_service.rb:244]: Ah, I see it now.  Small quibble - we're reassigning the variable to itself after sanitizing it.  It doesn't really make a difference from a functional standpoint, but I think it's safer to rename the sanitized array something else.

File: test/js/explore/plot-data-cache.test.js
Comment (bistline) [inline test/js/explore/plot-data-cache.test.js:85]: What does the `230` map to?  Is that the `perfTime` calculation?

**Review by bistline — CHANGES_REQUESTED** [review]
This looks good, but there is the blocking bug with my suggestion for cleaning up `ClusterVizService.load_cluster_group_data_array_points`.  Also, I have noticed an issue when testing locally.  If you load a cluster/annotation that has blank values (either from an unmapped cell, or a blank entry in a metadata file), then the scatter plot looks off:

Screen Shot 2021-05-18 at 10 17 35 AM

As you can see, the points aren't the same size as all the rest (from the study default options), and the DNA spinner never goes away.  Definitely a corner case, as we don't have that many of these in the portal, but something that should be looked at and addressed.

File: app/javascript/components/explore/plotDataCache.js
Comment (eweitz) [reply app/javascript/components/explore/plotDataCache.js]: Thanks for the insightful chat.  In summary:

1.  We agree that omitting props for perfTime segments that don't happen can be accounted for, by filtering analytics queries with logged props like `cache: "client/server/none"`.  That approach makes analysis harder (but not impossible) and simplifies instrumentation, so it's net reasonable.

2.  We'll need to be careful that omitted props don't skew summary metrics, e.g. definitely not `perfTimeFull`, and ideally not `perfTime:frontend` or `perfTime:backend`.

I'll re-review once those takeaways are implemented here or ticketed for this sprint.

**Review by eweitz — COMMENTED** [review]
Thanks for the insightful chat on [zero vs. undefined](https://github.com/broadinstitute/single_cell_portal_core/pull/1036#discussion_r633822328). In summary:

We agree that omitting props for `perfTime` segments that don't happen can be accounted for when analyzing within a segment (e.g. "How has transfer performance changed over time?") and across segments ("What's the principal bottleneck in performance?"), by filtering analytics queries with logged props like `cache: "client/server/none"`.  That approach makes analysis harder (but not impossible) and simplifies instrumentation, so it's net reasonable.

We'll need to be careful that omitted props don't skew summary metrics, e.g. definitely not `perfTime:full`, and ideally not `perfTime:frontend` or `perfTime:backend`.  For example, we'll want to ensure that when transfers don't happen (due to caching), we properly account for that in `perfTime:full` and ideally e.g. `perfTime:frontend`.

I'll re-review once those takeaways are implemented here or ticketed for this sprint.

File: test/js/explore/plot-data-cache.test.js
Comment (devonbush) [reply test/js/explore/plot-data-cache.test.js:85]: yeah, so for purposes of this test it just needs to be any number/object

File: app/lib/cluster_viz_service.rb
Comment (devonbush) [reply app/lib/cluster_viz_service.rb:244]: Good point, since this is at the end of the function, we actually don't need a variable for it at all

File: app/lib/cluster_viz_service.rb
Comment (devonbush) [reply app/lib/cluster_viz_service.rb:194]: good catch -- I've just renamed it to concatenate_data_arrays for consistency

**Comment by devonbush** [comment]
@bistline  I can't reproduce the missing point issue -- I can render clusters with 'unspecified' cells ok.  I wonder if there's something specific about the structure in the cluster you're using that brings out the bug.  Let's screenshare tomorrow -- since the spinner is still going, chances are good there'll be an informative error in the JS console

**Review by bistline — APPROVED** [review]

**Review by eweitz — APPROVED** [review]

**Pull request merged**
